network/transport/v2/scheduler.go

```diff
@@ -110,9 +114,23 @@ func (p *payloadScheduler) Run() error {
 	})
 }
 
+func (p *payloadScheduler) addFailedJob(hash hash.SHA256Hash) {
+	p.failedJobsLock.Lock()
+	defer p.failedJobsLock.Unlock()
+
+	p.failedJobs = append(p.failedJobs, hash)
+}
+
+func (p *payloadScheduler) GetFailedJobs() []hash.SHA256Hash {
+	p.failedJobsLock.RLock()
+	defer p.failedJobsLock.RUnlock()
+
+	return p.failedJobs
```

Comment:
access is still racy: when a `tx` is added after `GetFailedJobs` is called and returned: the diagnostics handler might be using the slice (e.g. marshalling to JSON) while another entry is being added.

For all diagnostics applies: best to work on copies of mutable data structures. Easiest is to make `GetFailedJobs()` return a copy. Then caller also can't mutate the internal state by altering the slice (although theoretically you can still alter the hashes inside it, since `hash.SHA256Hash` is just a slice)

Reply:
If you query the DB on all jobs that have a retry count equal to max wouldn't that give you the failed jobs?

Reply:
It most definitely would. Also makes for a more solid implementation.